pkg/adapter/adapter.go

```diff
@@ -31,17 +35,127 @@ type envConfig struct {
-// then sends them to the specified Sink
+// then sends them to the specified Sink.
 func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client) adapter.Adapter {
-	fmt.Printf("Hey from mongodb adapter")
-	time.Sleep(600 * time.Second)
+	env := processed.(*envConfig)
+
+	rawURI, err := ioutil.ReadFile(env.MongoDbCredentialsPath + "/URI")
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Unable to get MongoDb URI field", zap.Any("secretPath", env.MongoDbCredentialsPath+"/URI"))
+	}
+	URI := string(rawURI)
+
+	return newAdapter(ctx, env, ceClient, URI)
+}
+
+func newAdapter(ctx context.Context, env *envConfig, ceClient cloudevents.Client, URI string) adapter.Adapter {
+	// Create new Client.
+	client, err := mongo.NewClient(options.Client().ApplyURI(URI))
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Error creating mongo client", zap.Error(err))
+	}
+
+	var collection *mongo.Collection = nil
+	if env.Collection != "" {
+		collection = client.Database(env.Database).Collection(env.Collection)
+	}
+
+	return &mongoDbAdapter{
+		namespace: env.Namespace,
+		ce:        ceClient,
+		client:    client,
+		db:        client.Database(env.Database),
+		coll:      collection,
+		source:    env.EventSource,
+		sinkurl:   env.SinkURL,
+	}
+}
+
+// Start creates the watch stream that will watch for dataSource changes.
+func (a *mongoDbAdapter) Start(ctx context.Context) error {
+	// Connect to Client.
+	err := a.client.Connect(ctx)
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Error connecting to database", zap.Error(err))
+	}
+	defer a.client.Disconnect(ctx)
+
+	// Create a watch stream for either the database or collection.
+	var stream *mongo.ChangeStream = nil
+	if a.coll != nil {
+		stream, err = a.coll.Watch(ctx, mongo.Pipeline{})
+	} else {
+		stream, err = a.db.Watch(ctx, mongo.Pipeline{})
+	}
+	if err != nil {
+		return err
+	}
+	defer stream.Close(ctx)
+
+	// Watch and process changes.
+	a.processChanges(ctx, stream)
 	return nil
 }
+
+// processChanges processes the new incoming change, creates a cloud event and sends it.
+func (a *mongoDbAdapter) processChanges(ctx context.Context, stream *mongo.ChangeStream) {
+	// For each new change recorded.
+	for stream.Next(ctx) {
+		var data bson.M
+		if err := stream.Decode(&data); err != nil {
+			panic(err)
+		}
+		// Create corresponding event.
+		event, err := makeCloudEvent(data)
+		if err != nil {
+			logging.FromContext(ctx).Desugar().Error("Failed to create event", zap.Error(err))
+		}
+
+		// Send event using client.
+		ctx := cloudevents.ContextWithTarget(ctx, a.sinkurl)
+
+		// Send that Event.
+		if result := a.ce.Send(ctx, *event); cloudevents.IsUndelivered(result) {
+			logging.FromContext(ctx).Desugar().Error("Failed to send event")
+		}
+	}
+}
+
+// makeCloudEvent makes a cloud event out of the change object recevied.
+func makeCloudEvent(data bson.M) (*cloudevents.Event, error) {
```

Comment:
Doesn't look like this will ever return err?

Reply:
True will add error returns